Async handler approach

Krulknul opened the pull request:
PR against #2, to show what is needed to introduce full async functionality.
Includes some unrelated things as well like cleanup and a channel stream

Krulknul committed:
async implementation working

Krulknul committed:
improve macros and change naming

Krulknul committed:
introduce channel stream

Krulknul committed:
improve channel example

Krulknul committed:
refactor processor a bit

Krulknul committed:
make clippy a bit happier

Krulknul committed:
improve naming and some docs

Krulknul committed:
improve naming in transactionhandlercontext

Krulknul committed:
update one comment

Krulknul committed:
improve some comments

Krulknul committed:
remove a double comment

Krulknul committed:
add comment about temporary bug

Krulknul committed:
update cargo.lock

fpieper commented on examples/minimal.rs:
```diff
@@ -1,9 +1,8 @@
-use ::auto_decode::auto_decode;
+use handler_macro::event_handler;
```
Is there a reason this is a separate package?

I like the name of the macro but probably would expose it from the `radix_event_stream` package.
You could have something like `radix_event_stream::macros::event_handler`. If there is something against it you might also want to expose the important things at the top level like

`radix_event_stream::event_handler`

Krulknul replied:
There is a technical reason for this:

[rust book](https://doc.rust-lang.org/book/ch19-06-macros.html#procedural-macros-for-generating-code-from-attributes)

_"When creating procedural macros, the definitions must reside in their own crate with a special crate type. This is for complex technical reasons that we hope to eliminate in the future."_

Regarding `radix_event_stream::event_handler`, we already have an `event_handler` module so I would indeed do `radix_event_stream::macros::event_handler`, which is nicer, yes.

fpieper commented on examples/minimal.rs:
```diff
@@ -24,7 +23,7 @@ pub struct InstantiateEvent {
 }
 
-#[auto_decode]
+#[event_handler]
```
I like the name 👍

fpieper commented on src/event_handler.rs:
```diff
@@ -105,8 +88,8 @@ where
-    pub transaction: &'a IncomingTransaction,
+    pub incoming_transaction: &'a IncomingTransaction,
+    pub incoming_event: &'a IncomingEvent,
     pub transaction_handle: &'a mut TRANSACTION_HANDLE,
```
I guess we should have better names here.

Maybe better use `transaction_context` instead of `transaction_handle`

Krulknul replied:
Yep, this is something I was struggling with.

I wanted to name it transaction_context, but I didn't want it to be confused with the `TransactionHandlerContext`, which is the struct that is passed to transaction handlers.

Krulknul replied:
transaction handlers receive a TransactionHandlerContext
event handlers receive an EventHandlerContext

Now we also want to have a TransactionContext passed to event handlers, which would be a very logical name, but easy to confuse with the TransactionHandlerContext which is a very different thing.

Krulknul replied:
let's name it TransactionContext

fpieper commented on src/event_handler.rs:
```diff
@@ -105,8 +88,8 @@ where
 {
     pub app_state: &'a mut STATE,
-    pub transaction: &'a IncomingTransaction,
+    pub incoming_transaction: &'a IncomingTransaction,
```
Maybe instead of `incoming` we specify more clearly what it is e.g.:

`ledger_transaction` or anything along the lines of that?

Though `transaction_context` is probably already good enough to distinguish. Let me know what you think 👍

Krulknul replied:
I think LedgerTransaction would indeed be a better name than IncomingTransaction

fpieper replied:
I think I would go with `LedgerTransaction`. Maybe a bit convoluted but very explicit 👍

fpieper commented on examples/pools/main.rs:
```diff
@@ -61,40 +57,34 @@ fn main() {
         Ok(())
     }
 
     if arg == "file" {
```
In general, better to use a match here for the different cases and have each case in its own function

fpieper commented on examples/pools/main.rs:
```diff
@@ -61,40 +57,34 @@ fn main() {
-        )?;
+        context
+            .incoming_transaction
+            .handle_events(
```
I like the idea of abstracting away the loop 👍 . Maybe rename it to `process_events`?

Krulknul replied:
Not really a strong opinion on the naming here.

fpieper replied:
Let's go with `process_events` then. We already have enough `hanlders`^^

fpieper reviewed (approved):
Really great work @Krulknul 👍  Thank you.

I left some minor comments mostly about style and naming. We can also have a quick call to discuss them if you prefer that.

Krulknul committed:
expose macros in `macros` module

Krulknul committed:
clean up pools example

Krulknul commented:
> Really great work @Krulknul 👍 Thank you.
> 
> I left some minor comments mostly about style and naming. We can also have a quick call to discuss them if you prefer that.

Thanks for the review!

Krulknul committed:
update readme

Krulknul committed:
update readme

Krulknul committed:
update readme

Krulknul committed:
update readme

Krulknul committed:
remove pub in readme

Krulknul committed:
Add more info section to README.md

Krulknul committed:
changes as discussed:
change TRANSACTION_HANDLE to TRANSACTION_CONTEXT
Remove "Incoming" from ledger transaction and event models
rename
Handle decoding errors more gracefully in macro
Update examples
Update/improve readme

Krulknul committed:
Increment version to 1.0.0

Krulknul committed:
Update minimal and channel example
